returnn/tf/util/basic.py

```diff
@@ -1339,6 +1340,142 @@ def __call__(self, shape, dtype=None, partition_info=None):
+  ICASSP 2007 and also
+  https://en.wikipedia.org/wiki/Greenwood_function
+  The default values are taken from RASR, see https://github.com/rwth-i6/rasr/blob/master/src/Signal/GammaTone.cc
+  :param float freq:
+  :param float scaling_constant: A in [1]
+  :param float constant_of_integration: k in [1]
+  :param float slope: a in [1]
+  :return: fractional length corresponding to given frequency
+  :rtype: float
+  """
+  return np.log10(freq / scaling_constant + constant_of_integration) / slope
+
+
+class GammatoneFilterbankInitializer(init_ops.Initializer):
+  """
+  Initializer for a gammatone filterbank, e.g., to initialize weights of a convolutional layer.
+  Based on
+  [1] Schlueter, Ralf, et al. "Gammatone features and feature combination for large vocabulary speech recognition."
+  ICASSP 2007
+  """
+  def __init__(self, num_channels, length, sample_rate=16000, freq_max=7500., freq_min=100., normalization=True):
+    """
+    :param int num_channels: number of filters
+    :param int length: length of FIR filters in seconds
+    :param int sample_rate: sample rate of audio signal in Hz
+    :param float freq_max: maximum frequency of filterbank
+    :param float freq_min: minimum frequency of filterbank
+    :param bool normalization: normalize filterbanks to maximum frequency response of 0 dB
+    """
+    self.num_channels = num_channels
+    self.length = length
+    self.sample_rate = sample_rate
+    self.freq_max = freq_max
+    self.freq_min = freq_min
+    self.normalization = normalization
+
+  def __call__(self, shape, dtype=None, partition_info=None):
+    """
+    :param tuple[int] shape:
+    :param tf.DType dtype:
+    :param partition_info:
+    :rtype: tf.Tensor
+    """
+    center_freqs = self.center_frequencies(self.num_channels, self.freq_max, self.freq_min)
+    fbank = []
+    for freq in center_freqs:
+      fbank.append(self.gammatone_impulse_response(freq, self.length, self.sample_rate))
+    fbank = np.vstack(fbank)
+    if self.normalization:
+      fbank = self.normalize_filters(fbank)
+    fbank = np.reshape(np.transpose(fbank), shape).astype(dtype.as_numpy_dtype)
+    return tf.convert_to_tensor(fbank, dtype=dtype)
+
+  @staticmethod
+  def center_frequencies(num_channels, freq_max, freq_min):
+    """
+    Determine center frequencies for gammatone filterbank
+    :param int num_channels: number of filters
+    :param float freq_max: maximum frequency of filterbank
+    :param float freq_min: minimum frequency of filterbank
+    :return: center frequencies
+    :rtype: np.array
+    """
+    x_min = inv_greenwood_function(freq_min)
+    x_max = inv_greenwood_function(freq_max)
+    x_linear = np.linspace(x_min, x_max, num_channels)
+    freq_center = np.vectorize(greenwood_function)(x_linear)
+    return freq_center
+
+  @staticmethod
+  def bandwidth_by_center_frequency(freq, l_=24.7, q_=9.264491981582191):
```

Comment:
Why do you use `l_` or `q_` as name? Why the postfix `_`?

Reply:
I took the names from RASR and kept the postfix because PyCharm issues a warning if a variable is just named `l`. I could rename `q_` into `quality_factor`, but for `l_` I don't know a good name. I checked the paper and couldn't find a descriptive name in there.